scripts/opentitan/verilate.py

```diff
@@ -0,0 +1,593 @@
+from fcntl import fcntl, F_GETFL, F_SETFL
+from io import BufferedRandom, TextIOWrapper
+from os import O_NONBLOCK, getcwd, getenv, rename, symlink, unlink, walk
+from os.path import (basename, dirname, exists, isdir, isfile, islink,
+                     join as joinpath, normpath, realpath, splitext)
+from select import POLLIN, POLLERR, POLLHUP, poll as sel_poll
+from shutil import copyfile, rmtree
+from subprocess import Popen, PIPE, TimeoutExpired
+from sys import exit as sysexit, modules, stderr
+from tempfile import mkdtemp, mkstemp
+from threading import Thread
+from time import sleep, time as now
+from traceback import format_exc
+from typing import Optional, Union
+
+import logging
+import re
+import sys
+
+QEMU_PYPATH = joinpath(dirname(dirname(dirname(normpath(__file__)))),
+                       'python', 'qemu')
+sys.path.append(QEMU_PYPATH)
+
+# pylint: disable=wrong-import-position
+# pylint: disable=wrong-import-order
+# pylint: disable=import-error
+
+getLogger = logging.getLogger
+
+from ot.util.argparse import ArgumentParser, FileType  # noqa: E402
+from ot.util.file import guess_file_type, make_vmem_from_elf
+from ot.util.log import ColorLogFormatter, configure_loggers
+
+VeriVcpDescriptor = tuple[str, BufferedRandom, bytearray, logging.Logger]
+
+DEFAULT_TIMEOUT = 10.0 * 60
+"""Default execution timeout."""
+
+
+class VtorFileManager:
+    """File manager.
+
+       Handle temporary directories and files.
+
+       :param keep_temp: do not automatically discard generated files on exit
+       :param tmp_dir: store the temporary files in the specified directory
+    """
+
+    def __init__(self, keep_temp: bool = False, tmp_dir: Optional[str] = None):
+        self._log = getLogger('vtor.file')
+        self._keep_temp = keep_temp
+        self._base_tmp_dir = tmp_dir
+        self._tmp_dir: Optional[str] = None
+        register(self._cleanup)
+
+    @property
+    def tmp_dir(self) -> str:
+        """Provides the main temporary directory for executing Verilator."""
+        if not self._tmp_dir:
+            self._tmp_dir = mkdtemp(prefix='verilator_ot_dir_',
+                                    dir=self._base_tmp_dir)
+        return self._tmp_dir
+
+    def _cleanup(self):
+        if self._tmp_dir and isdir(self._tmp_dir):
+            if self._keep_temp:
+                self._log.warning('Preserving temp dir %s', self._tmp_dir)
+                return
+            self._log.debug('Removing temp dir %s', self._tmp_dir)
+            rmtree(self._tmp_dir)
+            self._tmp_dir = None
+
+
+class VtorExecuter:
+    """Verilator wrapper.
+
+        -c|--term-after-cycles=N
+    """
+
+    VMEM_OFFSET = 0
+    """Offset when converting BM test to VMEM file."""
+
+    ANSI_CRE = re.compile(rb'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]')
+    """ANSI escape sequences."""
+
+    DV_CRE = re.compile(r'^(\d+):\s\(../.*?\)\s(.*)$')
+    """DV macro messages."""
+
+    DEADLOCK = 125
+    """Default error code w/ Verilator is stuck."""
```

Comment:
Aside: maybe this should be:


```suggestion
    """Default error code when Verilator is stuck."""
```

(I have only ever seen `w/` as a shorthand for "with", but maybe this is just something I have never seen!)